feat(api, shared-data): Move lid command implementation

CaseyBatten opened the pull request:
# Overview
Covers EXEC-1005

Introduce `move_lid()` command to PAPI. Allows the movement of lids from stacks and source labware to destination labwares, stacks, or trash location. When a lid stack is emptied, it is moved to the new "systemLocation" location.

## Test Plan and Hands on Testing

- [x] The following protocol should pass analysis and execute on a robot:

```
from typing import List, Dict, Any, Optional
from opentrons.protocol_api import ProtocolContext, Labware

metadata = {"protocolName": "Opentrons Flex Deck Riser with TC Lids Test"}
requirements = {"robotType": "Flex", "apiLevel": "2.23"}

LID_DEFINITION = "opentrons_tough_pcr_auto_sealing_lid"
DECK_RISER_NAME = "opentrons_flex_deck_riser"
USING_THERMOCYCLER = True

def run(protocol: ProtocolContext):
    # SETUP
    deck_riser_adapter = protocol.load_adapter(DECK_RISER_NAME, "A2")
    trash = protocol.load_trash_bin("A3")

    stack = protocol.load_lid_stack(LID_DEFINITION, "B2", 5)
    
    thermocycler = protocol.load_module("thermocyclerModuleV2")
    thermocycler.open_lid()
    plate_in_cycler = thermocycler.load_labware("armadillo_96_wellplate_200ul_pcr_full_skirt")


    protocol.move_lid(stack, plate_in_cycler, True)

    # Move the lid to the empty riser, creating a new lid stack
    used_stack = protocol.move_lid(plate_in_cycler, deck_riser_adapter, True)
    protocol.move_lid(stack, plate_in_cycler, True)
    protocol.move_lid(plate_in_cycler, used_stack, True)
    protocol.move_lid(stack, plate_in_cycler, True)
    protocol.move_lid(plate_in_cycler, used_stack, True)
    protocol.move_lid(stack, plate_in_cycler, True)
    protocol.move_lid(plate_in_cycler, used_stack, True)
    protocol.move_lid(stack, plate_in_cycler, True)
    protocol.move_lid(plate_in_cycler, used_stack, True)

    # Trash all the lids
    protocol.move_lid(used_stack, trash, True)
    protocol.move_lid(used_stack, trash, True)
    protocol.move_lid(used_stack, trash, True)
    protocol.move_lid(used_stack, trash, True)
    protocol.move_lid(used_stack, trash, True)

```

## Changelog
Introduces new move_lid() PAPI commands, alongside new state update commands to allow for lid updating and lid stack invalidation.

## Review requests
Does the approach to returning Labware or None for move lid make sense when we're dealing with creating a new lid stack by moving a lid? The thought process here is:

1. I have a lid stack in A1
2. I move a lid from the lid stack in A1 to an empty deck slot in A2
3. I have de facto created a new lid stack in A2, so I should create a stack there and return it to the user for handling
4. If no new stack has been created at the lid's new position, I should return Nothing

## Risk assessment
Low, new behavior and it gated behind API 2.23. Also impossible to use on earlier API versions anyways given the lack of Lid labware IDs on parent labwares, lack of lid stacks, and lack of the optional lid field on labwares.

CaseyBatten committed:
move lid behavior implementation

CaseyBatten committed:
move lid behavior and stack return implementation

CaseyBatten committed:
labware definition updates

CaseyBatten committed:
move lid labware invalidation and deck slot support checking

sfoster1 reviewed (approved):
This makes sense to me, yeah. This is a lot of code that's now in both move labware and move lid, can we factor some of it into something common?

CaseyBatten committed:
schema updates and test corrections

CaseyBatten committed:
correct definition version

CaseyBatten reviewed (commented):
Note: Refactoring based on discussion regarding client and PD needs for load lid stack -> move labware -> move lid -> move labware command execution stackup.

CaseyBatten committed:
transition to api core related movement command and deprecation of engine move lid

CaseyBatten committed:
removal of move lid engine command

CaseyBatten committed:
Merge branch 'edge' into move_lid_command_implementation

CaseyBatten committed:
step generator move labware cleanup

CaseyBatten committed:
further js cleanup for systemLocation

CaseyBatten committed:
addition of biorad plate to TC lid

CaseyBatten committed:
ignore system location check in unoccupied trash

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal move_lid_command_implementation snapshots (#17299)

This PR was requested on the PRhttps://github.com/Opentrons/opentrons/pull/17259
Update snapshot tests to reflect TC lid compatible parents
